### ISSUE TITLE
Multiple instance producers per type

### ISSUE THREAD

**Issue by zedr0n** [simpleinjector/SimpleInjector#193]
Is it possible to have multiple registrations for the same type which get invoked depending on some context information? This is same as #171 but in a much more static context without WPF/MVVM.

E.g, I currently have properties

``` c#    
[Selector("input#email")]
protected override WebInputElement Username { get; set; }
[Selector("input#password")]
protected override WebInputElement Password { get; set; }
```

with constructor

``` c#
WebInputElement(IElementProvider elementProvider, string selector) 
```

I've used parameter convention using IDependencyInjectionBehaviour to override the expression built for selector depending on the attribute but the expression is built once per type so when I build password it instead creates the same instance as for Username. Would be nice to be able to extend container to be able to differentiate between types enriched with attributes maybe? ( at the moment, looking at the source the instances are built via a generic < , > which means that all attribute information is lost and only preserved within InjectionConsumerInfo )

For the moment I'm using a very scary hack by creating generic derived type WebInputElement< T > and then using a type created at runtime based on the attributes hashcode. 

``` c#
container.RegisterConditional(typeof(WebInputElement),
	c => typeof(WebInputElement<>).MakeGenericType(DynamicType.CreateType(c.Consumer.Target)),
	Lifestyle.Transient,
	c => c.Consumer != null && c.Consumer.ImplementationType.Is<WebContainer>());
```

I can see another way to also handle this, by explicitly modifying the built expression using IdependencyInjectionBehaviour when creating the instance of the class containing the properties instead of when the actual primitive string is built. But I'm not so familiar with expression trees and not sure if there's a clean way to do this. Would this be a better way?

Registering a collection instead doesn't seem to be possible as actual implementation type is still the same, it's actually the constructed instance should just have a different parameter injected.

I could probably again change primitives to provider services and use scoping but this is not really runtime, this is pure static graph initialisation.

For UWP Reflection.Emit is not available so I actually have to use MakeArrayType with number of dimensions based on attribute hashcode. This is really a bit hardcore for what I want to achieve :)

``` c#
public static class DynamicType
{
	private static readonly Dictionary<Tuple<Type,int>, int> Hash = 
		new Dictionary<Tuple<Type,int>, int>();
	private static readonly Dictionary<Type,int> MaxHash = new Dictionary<Type, int>();

	private static Type CreateType(Type type, int hashCode)
	{
		lock (Hash)
		{
			var tuple = new Tuple<Type, int>(type, hashCode);

			if (!Hash.ContainsKey(tuple))
			{
				if (!MaxHash.ContainsKey(type))
					MaxHash.Add(type, 1);
				else
					MaxHash[type]++;

				Hash.Add(tuple, MaxHash[type]);
			}
			else
				Hash[tuple] = MaxHash[type];

			return typeof(DynamicType).MakeArrayType(Hash[tuple]);
		}
	}

	public static Type CreateType(Type type, IEnumerable<Attribute> attributes)
	{
		var hashCode = attributes.Sum(attribute => attribute.GetHashCode());

		return CreateType(type, hashCode);
	}

	public static Type CreateType(InjectionTargetInfo info)
	{
		return CreateType(info.TargetType,info.Member?.GetCustomAttributes());
	}
}
```

For the situations where the registration is defined rather by the parents graph than attributes that can be incorporated in a cleaner way by using the consumer implementation type as a generic type parameter

``` c#
container.RegisterConditional(typeof(IElementProvider),
	c => typeof(GenericElementProvider<>).MakeGenericType(c.Consumer.ImplementationType),
	Lifestyle.Transient,
	c => c.Consumer != null);
```

giving at runtime a type like 

``` c#
GenericElementProvider<WebNumber<Profile>>
```

This guarantees that apart from property injection at least all constructor injected parameters are going to be re-resolved per each path in the graph.

In real code I actually combine both to create 

``` c#
IService<c.Consumer.ImplementationType,DynamicType.CreateType(c.Consumer.Target.Info.Member.GetCustomAttributes())>
```

**Comment by qujck** [comment]
I've read and reread this all evening and I'm sorry to say I cannot figure out what you are trying to do :-(

Do you have multiple `WebInputElement`'s, or you want a different `string` parameter injected into the same `WebInputElement` depending on a property attribute, or something else?

> Is it possible to have multiple registrations for the same type which get invoked depending on some context information?

Yes, register a collection and define a `CompositeWebInputElement` to sit over them to decide which `WebInputElement` is relevant.

> Registering a collection instead doesn't seem to be possible as actual implementation type is still the same, it's actually the constructed instance should just have a different parameter injected.

Can you make the parameter a run time value passed in by the `CompositeWebInputElement`

> `string selector`

Smells like a runtime parameter

**Comment by zedr0n** [comment]
public class Login : LoginService
    {        
        [Selector("input#email")]
        protected override WebInputElement Username { get; set; }
        [Selector("input#password")]
        protected override WebInputElement Password { get; set; }
     }
What I want is for this class instance to be constructed by the container. The target goal is that for Username WebInputElement build expression used by the container is smth like 
`
new WebInputElement(IElementProvider,"input#email")
`
and for Password is 
`
new WebInputElement(IElementProvider,"input#password")
` 
So I do have the same WebInputElement type and I want a different string injected into the parameter based on property attribute.

This can be achieved by overriding IDependencyInjectionBehaviour for the selector parameter of WebInputElement which basically just means that the constructor will have the value of the selector attribute injected.

The issue here is that the container resolves the Expression only once per type. So the first time the container will try to build WebInputElement it will find the selector attribute and replace the selector part of WebInputElement build expression with Expression.Constant("input#email"), giving 
`
new WebInputElement(IElementProvider,"input#email")
`

Now as container will start injecting the Password property it will see it's a WebInputElement, figure out it already has resolved the build expression for it and will reuse the expression without applying IDependencyInjectionBehaviour override again which means that for Password I will get again the same

`
new WebInputElement(IElementProvider,"input#email")
`

What I was looking for is a way to force the container to rebuild the expression again for Password.

I can obviously also manually create all the properties in the constructor like this

        public Login(IElementProvider provider)
        {
            Username = new WebInputElement(provider, "input#email");
            Password = new WebInputElement(provider, "input#password");
        }

but the whole point is that these are strings known at compile time so this should be possible to be resolved by the injector instead of me explicitly initialising properties in the constructor.

 Also it's simple to replace in this case but the actual property can depend on multiple attributes at different levels, e.g. the IElementProvider implementation type at some point in its construction depends on a url string which can be resolved from the url attribute higher in the construction graph. 

Before I was just using reflection to manually initialise all the properties based on the attributes but I like the idea of DI and decided to see if the same can be done with IoC instead.

**Comment by dotnetjunkie** [comment]
The dependency injection behavior is unsuited for this. You can use `RegisterConditional` because its predicate allows access to the injected member name, as in: `c => c.Consumer.Member.Name` (if I recall correctly from memory).

But to be honest, I think you are stretching this too much. I don't think you should be using the container's auto-wiring abilities for this. Instead, register a delegare to post-initialize your type:

``` C#
container.RegisterInitializer<Login>(login => {
    login.UserName = new WebInputElement(IElementProvider,"input#email");
   login.Password = new WebInputElement(IElementProvider,"input#password");
});
```

**Comment by zedr0n** [comment]
Using a post-initialize delegate I would have to explicitly write one for each single class which kind of defeats the purpose, it's easier and cleaner to put it into actual constructor then. 

Also IElementProvider by itself is a provider proxy depending on a scoped elementProvider and transient webelementprovider which depends on a url provider which is either a string equal to the Login class url attribute or a scoped value if it's constructed via abstract factory as in the other issue I've raised. Putting all this logic into initializer instead means explicit hand-wiring which kinda defeats the purpose anyways.

Basically, for some services the actual build expression is different for each construction graph path with context information so maybe it's about extending the container to somehow know which paths in graph are equivalent and which are different in terms of resulting build expression. Taken to extreme, easiest would be just to have a lifestyle which doesn't cache the built expression at all. As it's the caching which causes problems same as in #171

**Comment by dotnetjunkie** [comment]
> RegisterConditional is the solution I used in the end but I need to somehow have a different type for each value of c.Consumer.Member.Name

After looking it up in the API docs, I realized the call is actually `c.Consumer.Target`. You can get all the metadata.you need from the Target property. This includes data such as property name, property type and any defined attributes.

The predicate of `RegisterConditional` is called once per consumer, wjich makes it probably the ideal candidate for what you want to achieve. It does mean though that you need multiple `RegisterConditional` registrations to achieve your goal, but you might be able to call this method in some sort of loop.

**Comment by zedr0n** [comment]
Sorry, i think I'm misunderstanding what you mean by registerConditional here. What I'm doing currently is calling this in a loop

            container.RegisterConditional(typeof(WebInputElement),
                c => typeof(WebInputElement<>).MakeGenericType(DynamicType.CreateType(c.Consumer.Target)),
                Lifestyle.Transient,
                c => c.Consumer != null && c.Consumer.ImplementationType.Is<WebContainer>());

Are you talking about a different overload here? Because I cannot use the same implementation type multiple times in registerConditional?

**Comment by dotnetjunkie** [comment]
Hi @zedr0n,

I'm sorry for all the confusion. I'm on holiday and never read the whole discussion completely. This made me miss the overall picture.

If the behavior you are describing is correct, there is a bug in the handling of the `IDependencyInjectionBehavior`, because the `BuildExpression` method should be called once per dependency, while supplying the correct target information. This would give you all the information you need to build up the correct onject graph using the attributes on the properties.

As far as I see, the IDependencyInjectionBehavior abstraction is the only way to do this right now.

What you are doing however seems quite odd to me. I would typically build such graph of user interface components seperately from the object graph of application components, but perhaps @TheBigRic has something useful to say about.

Nonetheless does the concept of using this way of building your object graph intrigues me, and this would have been possible, if it weren't for the bug you are encountering. Do note that I think that PostSharp not really solves anything here, because mixing PostSharp  aspects with DI is almost impossible and the same holds for unit testing.

@qujck, can you confirm the described behavior? Do we have an unfortunate bug here?

**Comment by zedr0n** [comment]
Reading #171 and looking at the source i would think that it's expected behaviour unless the implementation type is actually different for each consumer. Otherwise it's going just to use the cached delegate for the type.

**Comment by dotnetjunkie** [comment]
Ahh, now I start to understand what is happening here. The expression is cached per consumer type, but this is not completely accurate, since you inject the same type multiple times into the same consumer, and in that case you  want BuildExpression to be called again for each property (and the same holds for ctor arguments as well).

**Comment by zedr0n** [comment]
Yes, although I think it's actually cached per implementation type. I.e even if I create another webinputelement directly it's going to reuse the same cached expression. Which makes sense in the concept of register conditional, it's really about specifying a different concrete implementation type

**Comment by zedr0n** [comment]
Looking at the source at least for attributes it's possible to change PredicateContext.ImplementationType to be a Tuple<Type,int?> instead of Type to achieve a rebuild as caching logic is based on the ImplementationType property as far as I see.
Like [this](https://github.com/zedr0n/SimpleInjector/commit/3f1a8c9f0bf67f0d79c267b72a385c2e62c575b1)

**Comment by dotnetjunkie** [comment]
Yep, that's what I was thinking of.

**Comment by zedr0n** [comment]
Also I think this is generalizable to almost any static condition. The way I see it, there's one important thing I needed - having not only the information 1 level up ( via InjectionConsumerInfo ) but the whole path in the graph. By adding a recursive InjectionConsumerInfo parent information to the constructor I can achieve just that. 

Now if I replace Tuple with a generic 

      TargetTypeInfo(InjectionConsumerInfo consumerInfo, Type concreteType)

the TargetTypeInfo has all the information about the whole graph path up to current dependency being injected. So I can now use attribute information of the parents to decide whether a new instance producer needs to be created ( and thus BuildExpression will be triggered ). 

I might have broken something but [here's](https://github.com/zedr0n/SimpleInjector/commit/e764935165c4d6abf07b21243759dd5290b6494e) what I managed to do.

In general, I even think that the whole dictionary can be encapsulated with a specific Func<InjectionConsumerInfo,bool> which can be overloaded via container options so an end user can explicitly define which graph paths should trigger a new instance producer and which one should reuse cached ones.

**Comment by dotnetjunkie** [comment]
There is actually a good reason why Simple Injector doesn't allow getting information about the grand parents (the consumer of a consumer), and that's because it would break one crucial promise, which is that a singleton registrations are guaranteed to have at most one instance.

Say for instance you have two conditional registrations for abstraction IX (classes D and E, both singletons) that get injected based on their consumer's consumer. So let's say we want to inject D in case the grand parent is A, and otherwise inject E. This breaks down immediately when the their direct consumer (C) is a singleton as well (where C depends on IX). This can be visualized very effectively by showing the two dependency graphs: 

    Graph1: A -> C -> D
    Graph2: B -> C -> E

Since C is promised to be a singleton, the two object graphs here can't coexist, because it would require at least two instances of C to build these graphs. If we on the other hand kept the promise of having exactly one instance of C, it would mean that either D or E would never be used; which would of course be a serious bug as well.

Other DI libraries actually contain this flaw. Ninject for instance actually allows getting information about the complete object graph which causes the described incorrect  behavior without any warning whatsoever. Unity has someting similar where it allows you to create a child container and change the registrations within such child container before starting to resolve. If you then swap registration D with E in the child container, which dependency gets injected into C depends solely on timing. When C is already once constructed before D was swapped, you get D, otherwise C will depend on E (for ever).

This is not a theoretical problem actually. I've seen questions on codeplex and stackoverflow about the behavior of these containers by developers who hot burned by this.

Although you might argue that it is always possible to shoot yourself in the foot, this is not Simple Injector's strategy. We want to prevent developers from doing this by design and we are very keen on keeping our API that way, even if it would mean that losing a feature that might benefit some.

**Comment by zedr0n** [comment]
I see, that does sound troublesome. Considering that SimpleContainer has the verification feature, is it possible to just throw if a singleton is detected in a particular graph path?

**Comment by dotnetjunkie** [comment]
IMO, verify is not strong enough. Even without calling Verify, the container should check this, but to make the feature useful, you only want to prevent the registration when someone calls `Consumer.Consumer` or `Consumer.Consumer.Consumer` etc. This basically means we need to do runtime checks (flag calls to .Consumer) mixed with the knowledge of whether there's a singleton consumer in the graph. This is possible, but I'm not sure I want to stretch it that far. 

Reason is that I've been able to help anyone who actually wanted this behavior by giving feedback on their design. Improving the design has -up till now- always removed the need for this feature.

You on the other hand, are using Simple Injector in a way that it was not designed for. You came a long way, because of SI's flexible design, but I'm unsure whether this complication in code is justified for one user who is basically doing crazy stuff with our container :)

Don't get me wrong. I'm intrigued by what you are trying to accomplish and I would say that staying away from PostSharp is always a good idea (assumed that you focus on design of course), and I am dedicated to fix any bugs we have (as you proved we have), but adding such feature (which requires maintenance and support), while bring something I usually advice against doing, is probably not worth it for us.

But there might be multiple ways to skin a cat.

**Comment by zedr0n** [comment]
That's the beauty of open source, at least I can pay maintenance costs myself I really really want such a feature :) But yeah, I just had a feeling I'm breaking something with such changes so wanted to double check if there were plans on doing something in this direction.

**Comment by dotnetjunkie** [comment]
Haha, I think you misunderstand open source :). The majority of the costs is not in creating the pull request. 

The last two year I spend less than 5% of my time on actual programming (writing code and unit tests). And this includes the working on v3 which was a major undertaking. The other 95% of the time was spent with things like answering questions, reading the code, writing and maintaining documentation and other types of communication such as communicating with the rest of the team, other DI experts and developers of other OSS libraries (and frameworks such as .NET core). Don't forget that even when we would accept a PR, we will have to write the unit tests and documentation, keep everyting consistent with the rest of the code, and the extra complexity added will have a constant cost when we investigate issues and debug the library. We will have to answer questions about this feature as well. And we will have to do this after the sender of the PR is long gone.

Don't get me wrong, I love getting PR requests (even if we won't accept them directly because the sender didn't sign the [CLA](https://simpleinjector.org/contribute)). Those PRs are still valuable to us and are often a good starting point for a conversation. PRs are therefore highly appreciated. But even if you were willing to pay all the maintenance costs (which would at least mean becoming a dedicated member of the team and answer all questions about the feature), there is still a product vision and we are very conservative about adding features.

**Comment by zedr0n** [comment]
I didn't mean it exactly that way, I'm also involved in programming in my career so I totally understand that actual coding is unfortunately a quite small part of the whole project as soon as it gets big enough.

I just meant that I could fork the code and hack it to my heart's content :)

**Comment by dotnetjunkie** [comment]
I created issue #194 as work item for the caching bug with IDependencyInjectionBehavior.

**Comment by zedr0n** [comment]
Thanks, would be very interesting to see the exact design solution you'll have in mind for that.

About the singleton issue. In SimpleInjector design the expression is built from top down so one way to deal with the issue is rather than check at runtime where Consumer.Consumer is called it's easiest to stop passing passing the grandparent information as soon the as the graph hits any singleton. As the passing logic resides within SimpleInjector and not modifiable by the end user even if limited configuration by grandparents is allowed via an extension, the actual parent info will be missing so the code wouldn just not work with an explicit null reference exception if somebody doesn't check for the explicit availability of that info

**Comment by zedr0n** [comment]
One more question if you don't mind :) Right now, with caching done by implementation type it's guaranteed pretty much that the resulting graph wouldn't have completely equal expressions. Now if this is extended to account for context info, in theory this could result in a big number of redundant instance producers ( yes, this is exactly the problem I'm having right now, with the dumb attribute hash of the whole path I end up with dozens of copies ). This issue is smaller if you only have access to the immediate parent info but might it be beneficial to implement some kind of graph reduction after it's fully initialized, e.g. after verification?

**Comment by dotnetjunkie** [comment]
Can you show how you are creating and using these InstancePtoducer instances? It will highly depend on what you are actually doing.

**Comment by zedr0n** [comment]
What happens is that as I use cache based on all parents and grandparents attributes, any dependency which is below in the graph would end up with multiple producers. But as they are not affected by IdependencyInjectionBehaviour override the resulting expression would be the same for all of them. But that's an extreme scenario.

In general though if we have

A depending on B,C,D

where the injection is different depending on some context info related to A. But this is only true for B, C, and D are always the same. Now unless the cache is complex enough to handle the difference ( which would make it quite complicated ) we would end up with redundant producers for С and D. If you have a lot of dependencies for A this could snowball quickly. So I was thinking that maybe after the container is locked it could parse the whole graph and merge the producers with the same resulting expression

**Comment by dotnetjunkie** [comment]
It won't, and can't merge producers. Having two producers has a different effext than having one. Take for insrance two singleton producers for type A. This will eventually result in two insrances, because each producer gets its own cache. Since this os typically not whqt you want, SI has a diagnostic error for this called 'torn lifestyle'. But automatically merging this would go too far, because the user might actually want this behavior, or intended to do something different. An auto merge would be a sort of 'fail silently' which is something SI prevents.

So in your case, you should typically prevent duplicates yourself.

About memory, I have no idea what the extra amount of memory use would be in your application and whether it matters. For desktop and web apps, the increase would probably be irrelevant, but whether that is the case in a mobile app, I don't know.

But do know that customly created producers can go out of scope and can get garbage collected. This includes the expression tree they created and cached internally. But although you can reason about this, do measure it. The use of SI causes an increase of memory use.

Last note, in theory it's even possible to use SI to generate the delegates that allow creating the object graphs and after that simply let the container (with all producers and registrations) go out of scope.

**Comment by zedr0n** [comment]
Interestingly enough, it seems a similar idea is used by xUnit.net for writing data-driven tests where the values defined by attributes are injected for actual test execution e.g. [here](http://xunit.github.io/docs/getting-started-desktop.html)

    [Theory]
    [InlineData(3)]
    [InlineData(5)]
    [InlineData(6)]
    public void MyFirstTheory(int value)
    {
        Assert.True(IsOdd(value));
    }

I assume there's some also dependency injection going on under the hood...

**Comment by dotnetjunkie** [comment]
Hi @zedr0n,

I've been trying to reproduce the issue you're having with property injection and the `IDependencyInjectionBehavior` class, but I'm unable to reproduce the issue. The bug doesn't seem to exist, but perhaps I misinterpreted your issue.

This is the code I used to reproduce:

``` c#
[TestMethod]
public void Test()
{
    // Act
    var container = new Container();

    container.Options.DependencyInjectionBehavior =
        new DebugDependencyInjectionBehavior(container.Options.DependencyInjectionBehavior);
    container.Options.PropertySelectionBehavior = 
        new PropertySelectionBehavior<SelectorAttribute>();

    container.GetInstance<Login>();
    container.GetInstance<Login>();
}

public class Login
{
    public Login(
        [Selector("input#acceptconditions")] WebInputElement conditions,
        [Selector("input#subscribe2newsletter")] WebInputElement newsLetter) { }

    [Selector("input#email")]
    public WebInputElement Username { get; set; }
    [Selector("input#password")]
    public WebInputElement Password { get; set; }
}

public class WebInputElement { }

private class DebugDependencyInjectionBehavior : IDependencyInjectionBehavior
{
    IDependencyInjectionBehavior real;

    public DebugDependencyInjectionBehavior(IDependencyInjectionBehavior real) {
        this.real = real;
    }

    public Expression BuildExpression(InjectionConsumerInfo consumer) {
        Debug.WriteLine(consumer.Target.GetCustomAttribute<SelectorAttribute>()?.Name);
        return this.real.BuildExpression(consumer);
    }

    public void Verify(InjectionConsumerInfo consumer) {
        this.real.Verify(consumer);
    }
}

private class PropertySelectionBehavior<TAttribute> : IPropertySelectionBehavior
    where TAttribute : Attribute
{
    public bool SelectProperty(Type serviceType, PropertyInfo propertyInfo) =>
        propertyInfo.GetCustomAttribute<TAttribute>() != null;
}

public class SelectorAttribute : Attribute
{
    public string Name { get; }
    public SelectorAttribute(string name) { this.Name = name; }
}
```

When I run the test with v3.1.2, I get the following output.

```
input#acceptconditions
input#subscribe2newsletter
input#email
input#password
```
This proves that both constructor arguments and properties aren't cached.

Is your issue different and did I misunderstand the issue, or did I mis something. Could you post some code that allows me to reproduce your issue?

**Comment by zedr0n** [comment]
In my case the attribute affects the arguments of WebInputElement constructor, not the parent consumer instance. So the problem is not in actual Login class instance which is not cached, but the  caching of the WebInputElement instances - i.e. in this case both WebInputElement properties actually have the same constructed object instead of being defined by selector.

I'll try to modify your code to demonstrate the actual problem when I get home
